File: leetcode/C++/reconstruct-itinerary.cpp
class Solution {
public:
    vector<string> findItinerary(vector<vector<string>>& tickets) {
        vector<string> sol={"JFK"};
        vector<int> visited(tickets.size(),0);
        stack<int> st;
        int s=0,t=-1;
        sort(tickets.begin(),tickets.end());
        while(st.size()<tickets.size()){
            s=0;
            for(int i=0; i<tickets.size(); ++i){
                if(visited[i] || i<=t) continue;
                if(tickets[i][0]==sol.back()){
                    visited[i]=s=1; t=-1;
                    sol.push_back(tickets[i][1]);
                    st.push(i);
                    break;
                }
            }
            if(s==0 && st.size()<tickets.size()) {
                t=st.top(); st.pop(); 
                visited[t]=0; sol.pop_back();
            }
        }
        return sol;
    }
};

File: leetcode/Python/reconstruct-itinerary.py
class Solution:
    def findItinerary(self, tickets: List[List[str]]) -> List[str]:
        sol, st, f, v, t = ["JFK"], [], 0, [0 for i in range(len(tickets))], -1
        tickets.sort()
        while len(st) < len(tickets):
            f=0
            for i in range(len(tickets)):
                if v[i]>0 or i<=t:
                    continue
                if tickets[i][0]==sol[-1]:
                    sol.append(tickets[i][1])
                    st.append(i)
                    t, f, v[i] = -1, 1, 1
                    break;
            if f==0:
                t=st.pop()
                sol.pop()
                v[t]=0
        return sol
                